### PR TITLE
Show `ContactDoctorSheet` when there are multiple teleconsultation phone numbers #2

### PR DESCRIPTION
**Pull request by msasikanth**
https://www.pivotaltracker.com/story/show/173240199

### CODE REVIEW

File: app/src/test/java/org/simple/clinic/summary/PatientSummaryUpdateTest.kt [hunk 1]
```diff
@@ -631,7 +631,7 @@ class PatientSummaryUpdateTest {
   }
 
   @Test
-  fun `when contact doctor button is clicked, then load patient teleconsultation information`() {
+  fun `when contact doctor button is clicked and only one phone number is present, then load patient teleconsultation information`() {
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/PatientSummaryUpdateTest.kt:634]: Instead of describing the implementation detail only one phone number is present, can we reword this to reflect the real world usage? Ex: `only one doctor is available for teleconsultation`.

File: app/src/test/java/org/simple/clinic/summary/PatientSummaryUpdateTest.kt [hunk 2]
```diff
@@ -655,6 +655,27 @@ class PatientSummaryUpdateTest {
         )
   }
 
+  @Test
+  fun `when contact doctor button is clicked and more than one phone number is present, then open contact doctor sheet`() {
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/summary/PatientSummaryUpdateTest.kt:659]: Instead of describing the implementation detail only one phone number is present, can we reword this to reflect the real world usage? Ex: `multiple doctors are available for teleconsultation`.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt [hunk 3]
```diff
@@ -135,6 +137,7 @@ class PatientSummaryScreen(
         .setActionTextColor(ContextCompat.getColor(context, R.color.green2))
         .setAnchorView(R.id.doneButtonFrame)
   }
+  private val detaches = detaches()
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:140]: Why was this extracted? It seems like it is not being used anywhere else.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt [hunk 4]
```diff
@@ -172,6 +175,20 @@ class PatientSummaryScreen(
     super.onDetachedFromWindow()
   }
 
+  private fun teleconsultPhoneNumberSelected(): Observable<PatientSummaryEvent> {
+    return screenRouter
+        .streamScreenResults()
+        .ofType<ActivityResult>()
+        .filter { it.requestCode == CONTACT_DOCTOR_SHEET && it.succeeded() && it.data != null }
+        .map { it.data!! }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:183]: We have this exact same codeblock everywhere where the only thing that changes is the type of the result data and the result code. Maybe we should think of moving this to an extension function.

```kotlin
inline fun <reified T> Observable<ActivityResult>.extractSuccessful(
    requestCode: Int,
    crossinline resultMapper: (Intent) -> T
): Observable<T> {
  return filter { it.requestCode == requestCode && it.succeeded() && it.data != null }
      .map { resultMapper(it.data!!) }
}
```

The usage then becomes a lot easier to read and developers can focus on the business logic instead of this boilerplate.

```kotlin
screenRouter
        .streamScreenResults()
        .ofType<ActivityResult>()
        .extractSuccessful(CONTACT_DOCTOR_SHEET, ContactDoctorSheet.Companion::readPhoneNumberExtra)
        .map { teleconsultPhoneNumberString ->
          val teleconsultPhoneNumber = TeleconsultPhoneNumber(teleconsultPhoneNumberString)

          ContactDoctorPhoneNumberSelected(teleconsultPhoneNumber)
        }
```

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt [hunk 5]
```diff
@@ -425,7 +442,11 @@ class PatientSummaryScreen(
       facility: Facility,
       phoneNumbers: List<TeleconsultPhoneNumber>
   ) {
-    // TODO (SM): Open contact doctor sheet
+    val phoneNumbersArrayList = arrayListOf<TeleconsultPhoneNumber>().apply {
+      addAll(phoneNumbers)
+    }
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:447]: Why are we creating a new list here of the exact same thing? We can pass the `phoneNumbers` list directly to the intent factory method, right?

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,8 +4,8 @@
 ### Feature
 - User can soft delete patient
 - Add diagnosis and blood sugar readings to the teleconsultation message
+- Support multiple teleconsultation phone numbers
```
Comment (vinaysshenoy) [inline CHANGELOG.md:7]: Can we reword this to better describe a user facing feature?

Ex: "User can now select from multiple doctors when requesting a teleconsultation".

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt [hunk 7]
```diff
@@ -99,20 +99,32 @@ class PatientSummaryUpdate : Update<PatientSummaryModel, PatientSummaryEvent, Pa
...
+          }
+          else -> {
+            LoadPatientTeleconsultationInfo(
+                model.patientUuid,
+                model.patientSummaryProfile?.bpPassport,
+                model.currentFacility,
+                teleconsultInfo.doctorsPhoneNumbers.first()
+            )
+          }
+        }
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:116]: I mainly did that for readability, but once I extract the boolean logic I can switch to an if/else condition.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt [hunk 8]
```diff
@@ -99,20 +99,32 @@ class PatientSummaryUpdate : Update<PatientSummaryModel, PatientSummaryEvent, Pa
...
-      is TeleconsultInfo.Fetched -> dispatch(LoadPatientTeleconsultationInfo(
-          model.patientUuid,
-          model.patientSummaryProfile?.bpPassport,
-          model.currentFacility,
-          model.teleconsultInfo.doctorsPhoneNumbers.first()
-      ) as PatientSummaryEffect)
+    return when (val teleconsultInfo = model.teleconsultInfo) {
+      is TeleconsultInfo.Fetched -> {
+        val effect = when {
+          teleconsultInfo.doctorsPhoneNumbers.size > 1 -> {
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:105]: Yeah, since we know this won't be triggered when there are no phone numbers we will have either 1 or more numbers states.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt [hunk 9]
```diff
@@ -99,20 +99,32 @@ class PatientSummaryUpdate : Update<PatientSummaryModel, PatientSummaryEvent, Pa
...
+            OpenContactDoctorSheet(teleconsultInfo.doctorsPhoneNumbers)
```
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:106]: Yeah, makes sense 👍

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 5 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:447]: No, intent takes in an `ArrayList` and doesn't accept the Kotlin `List`.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 4 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:183]: Should this be part of this PR?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt (hunk 9 above)
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:106]: One more thing I forgot to add: this commit breaks the app. The reason being that the method that opens the contact patient sheet is a no-op on the UI so clicking the button when there are no doctors available will result in a button-click that does nothing.

We should squash the first three commits together so that this change goes in as a single commit.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 3 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:140]: Oh, I forgot to add `takeUntil` for the `streamScreenResults` 😅

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 5 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:447]: Then this is a leaky abstraction since the factory method is exposing this detail. We should change the factory method to expect a `List<>` and then convert it to an array list inside that method.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:183]: Not necessary.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:105]: Done

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 5 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:447]: Yeah, changed it

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt (hunk 9 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:106]: I have squashed the commits, but when no doctors are available the button should go to disable state. So the user won't be able to press the button.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 4 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt]: Can we create a ticket for this?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt (hunk 9 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt]: Sorry, I meant when more than one doctor is available.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 4 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt]: Done: https://www.pivotaltracker.com/story/show/173282881

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 3 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:140]: This is not necessary since Mobius disposes the streams when the screen gets detached.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:140]: We need to manually dispose streams only when they are not part of a Mobius loop.

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt [hunk 10]
```diff
@@ -99,20 +99,39 @@ class PatientSummaryUpdate : Update<PatientSummaryModel, PatientSummaryEvent, Pa
...
+      }
+      else -> noChange()
+    }
+  }
+
+  private fun eventForContactDoctorClicked(
+      teleconsultInfo: TeleconsultInfo.Fetched,
+      model: PatientSummaryModel
+  ): PatientSummaryEffect {
+    return if (areMultipleDoctorsAvailable(teleconsultInfo)) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:115]: I don't really think this needs to be a function since it is only used in one place. Maybe we can convert it to a computed property on the `TeleconsultInfo.Fetched` class instead?

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt (hunk 3 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryScreen.kt:140]: Ah okay 👍

File: app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt (hunk 10 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/summary/PatientSummaryUpdate.kt:115]: Ah right, forgot about that. Done
